File: nocows.cpp

#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <unordered_set>
#include <cmath>
#include <string>
#include <map>
#include <numeric>
#include <unordered_map>

using namespace std;
typedef long double ld;
typedef long long ll;

const int MOD = 9901;
int dp[500][500];

int fm(int a)
{

    int tmp = a / MOD;
    return a - (MOD * tmp);
    // return a % MOD;
}

void solve(ll n, ll k)
{
    if (n <= 0)
        return;

    if (k <= 0)
        return;

    if (n % 2 == 0)
    {
        dp[n][k] = 0;
        return;
    }

    if ((n < (2 * k - 1)))
    {
        dp[n][k] = 0;
        return;
    }

    if (n > ((ll)pow(2, k) - 1) && k < 10)
    {
        dp[n][k] = 0;
        return;
    }

    // Both trees have height k - 1
    for (int ln = 1; ln < n; ln++)
    {
        int rn = n - ln - 1;
        int l = fm(dp[ln][k - 1]);
        int r = fm(dp[rn][k - 1]);
        dp[n][k] = fm((dp[n][k] + (fm(l * r))));
    }

    // Left tree has height k - 1
    for (int rh = 1; rh < k - 1; rh++)
    {
        for (int ln = 1; ln < n; ln++)
        {
            int rn = n - ln - 1;
            int l = fm(dp[ln][k - 1]);
            int r = fm(dp[rn][rh]);
            dp[n][k] = fm((dp[n][k] + (fm(l * r))));
        }
    }

    // Right tree has height k - 1
    for (int lh = 1; lh < k - 1; lh++)
    {
        for (int ln = 1; ln < n; ln++)
        {
            int rn = n - ln - 1;
            int l = fm(dp[ln][lh]);
            int r = fm(dp[rn][k - 1]);
            dp[n][k] = fm((dp[n][k] + (fm(l * r))));
        }
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    freopen("nocows.in", "r", stdin);
    freopen("nocows.out", "w", stdout);
    dp[1][1] = 1;
    int n, k;
    cin >> n >> k;
    for (int i = 3; i <= n; i += 2)
    {
        for (int j = 0; j <= k; j++)
        {
            solve(i, j);
        }
    }
    cout << dp[n][k] << "\n";
}

File: nocows.py


from sys import stdin, stdout

md = 9901

stdin = open("nocows.in", "r")
stdout = open("nocows.out", "w")

n, k = map(int, stdin.readline().strip().split())

dp = {}
dp[(1, 1)] = 1

def solve(n, k):
    if k <= 0:
        return 0
    
    if n <= 0 or n&1 == 0:
        return 0
    
    
    if n < (2*k - 1)  or n > (2**k) - 1:
        return 0
    
    if (n, k) in dp:
        return dp[(n, k)]
    
    ans = 0
    
    
    
    
    # both left and right have height k - 1
    for left_nodes in range(1, n, 2):
        right_nodes = n - left_nodes - 1
        l = (solve(left_nodes, k - 1)%md)
        r = (solve(right_nodes, k - 1)%md)
        ans = (ans + (l*r)%md)%md
        
        
    # left has height k - 1
    for right_height in range(1, k - 1):
        for left_nodes in range(1, n, 2):
            right_nodes = n - left_nodes - 1
            l = (solve(left_nodes, k - 1)%md)
            r = (solve(right_nodes, right_height)%md)
            ans = (ans + (l*r)%md)%md
            
            
    # # right has height k - 1
    for left_height in range(1, k - 1):
        for left_nodes in range(1, n, 2):
            right_nodes = n - left_nodes - 1
            l = solve(left_nodes, left_height)%md
            r = solve(right_nodes, k-1)%md
            ans = (ans + (l*r)%md)%md
            
            
    dp[(n, k)] = ans
    return dp[(n, k)]


stdout.write(f"{solve(n, k)}\n")
            
    
    
    
    
    